core/src/main/java/org/apache/iceberg/hadoop/Util.java

```diff
@@ -84,10 +88,38 @@ public static String[] blockLocations(FileIO io, ScanTaskGroup<?> taskGroup) {
     return locations.toArray(HadoopInputFile.NO_LOCATION_PREFERENCE);
   }
 
+  public static boolean usesHadoopFileIO(FileIO io, String location) {
+    if (io instanceof HadoopFileIO) {
+      return true;
+
+    } else if (io instanceof ResolvingFileIO) {
+      ResolvingFileIO resolvingFileIO = (ResolvingFileIO) io;
+      return resolvingFileIO.ioClass(location).isAssignableFrom(HadoopFileIO.class);
+
+    } else {
+      return false;
+    }
+  }
+
+  public static boolean mayHaveBlockLocations(FileIO io, String location) {
```

Comment:
I'd add the new methods after the `blockLocations` method below. The new code split `blockLocations` methods, which are related. It would be easier to have those next to each other. Otherwise, one would need to jump through the file.